http/src/v0/block.rs

```diff
@@ -98,31 +102,174 @@ pub fn put<T: IpfsTypes>(
+use std::fmt;
+
+fn rm_options() -> impl Filter<Extract = (RmOptions,), Error = Rejection> + Clone {
+    warp::filters::query::raw().and_then(|q: String| {
+        let res = RmOptions::try_from(q.as_str())
+            .map_err(StringError::from)
+            .map_err(warp::reject::custom);
+
+        futures::future::ready(res)
+    })
+}
+
+#[derive(Debug)]
+pub enum ParseError<'a> {
+    DuplicateField(Cow<'a, str>),
+    MissingArg,
+    InvalidNumber(Cow<'a, str>, Cow<'a, str>),
+    InvalidBoolean(Cow<'a, str>, Cow<'a, str>),
+}
+
+impl<'a> fmt::Display for ParseError<'a> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        use ParseError::*;
+        match *self {
+            DuplicateField(ref s) => write!(fmt, "field {:?} was duplicated", *s),
+            MissingArg => write!(fmt, "required field \"arg\" missing"),
+            InvalidNumber(ref k, ref v) => write!(fmt, "field {:?} invalid number: {:?}", *k, *v),
+            InvalidBoolean(ref k, ref v) => write!(fmt, "field {:?} invalid boolean: {:?}", *k, *v),
+        }
+    }
+}
+
+impl<'a> TryFrom<&'a str> for RmOptions {
+    type Error = ParseError<'a>;
+
+    fn try_from(q: &'a str) -> Result<Self, Self::Error> {
+        use ParseError::*;
+
+        let parse = url::form_urlencoded::parse(q.as_bytes());
+
+        let mut args = Vec::new();
+        let mut force = None;
+        let mut quiet = None;
+
+        for (key, value) in parse {
+            let target = match &*key {
+                "arg" => {
+                    args.push(value.into_owned());
+                    continue;
+                }
+                "force" => &mut force,
+                "quiet" => &mut quiet,
+                _ => {
+                    // ignore unknown fields
+                    continue;
+                }
+            };
+
+            // common bool field handling
+            if target.is_none() {
+                match value.parse::<bool>() {
+                    Ok(value) => *target = Some(value),
+                    Err(_) => return Err(InvalidBoolean(key, value)),
+                }
+            } else {
+                return Err(DuplicateField(key));
+            }
+        }
+
+        Ok(RmOptions {
+            args,
+            force: force.unwrap_or(false),
+            quiet: quiet.unwrap_or(false),
+        })
+    }
+}
+
+use futures::stream::FuturesOrdered;
+
+async fn rm_query<T: IpfsTypes>(
+    ipfs: Ipfs<T>,
+    options: RmOptions,
+) -> Result<impl Reply, Rejection> {
+    use futures::future::TryFutureExt;
+
+    let RmOptions { args, force, quiet } = options;
+
+    let futs: FuturesOrdered<_> = args
+        .into_iter()
+        .map(|hash| Cid::from_str(&hash).map_err(StringError::from).unwrap())
+        .map(|cid| ipfs.remove_block(cid.clone()).map_err(move |e| (cid, e)))
+        .collect();
+
+    let responses = futs
+        .collect::<Vec<Result<Cid, (Cid, Error)>>>()
+        .await
+        .into_iter()
```

Comment:
I still don't understand the collecting to a vec here, am I missing something why all the results need to be buffered?

Reply:
Tried this, got a `cannot infer an appropriate lifetime for autoref due to conflicting requirements` ... Probably because there's nothing keeping the `ipfs` alive until the end.